broadcaster/backlog/backlog.go

```diff
@@ -279,48 +265,72 @@ func newBacklogSegment() *backlogSegment {
 // variable of type BacklogSegment is nil or not. Comparing whether an
 // interface is nil directly will not work.
 func IsBacklogSegmentNil(segment BacklogSegment) bool {
-	return segment.(*backlogSegment) == nil
+	if segment == nil {
+		return true
+	} else if segment.(*backlogSegment) == nil {
+		return true
+	}
+	return false
 }
 
 // Start returns the first message index within the backlogSegment.
 func (s *backlogSegment) Start() uint64 {
-	return uint64(s.start.Load())
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	if len(s.messages) > 0 {
+		return uint64(s.messages[0].SequenceNumber)
+	}
+	return uint64(0)
 }
 
 // End returns the last message index within the backlogSegment.
 func (s *backlogSegment) End() uint64 {
-	return uint64(s.end.Load())
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
+	c := len(s.messages)
+	if c == 0 {
+		return uint64(0)
+	}
+	return uint64(s.messages[c-1].SequenceNumber)
 }
 
 // Next returns the next backlogSegment.
 func (s *backlogSegment) Next() BacklogSegment {
-	return s.nextSegment.Load()
+	next := s.nextSegment.Load()
+	if next == nil {
+		return nil // return a nil interface instead of a nil *backlogSegment
+	}
+	return next
 }
 
 // Messages returns all of the messages stored in the backlogSegment.
 func (s *backlogSegment) Messages() []*m.BroadcastFeedMessage {
 	s.messagesLock.RLock()
 	defer s.messagesLock.RUnlock()
-	return s.messages
+	tmp := make([]*m.BroadcastFeedMessage, len(s.messages))
+	copy(tmp, s.messages)
+	return tmp
 }
 
-// get reads messages from the given start to end message index.
-func (s *backlogSegment) get(start, end uint64) ([]*m.BroadcastFeedMessage, error) {
+// Get reads messages from the given start to end message index.
+func (s *backlogSegment) Get(start, end uint64) ([]*m.BroadcastFeedMessage, error) {
+	s.messagesLock.RLock()
+	defer s.messagesLock.RUnlock()
 	noMsgs := []*m.BroadcastFeedMessage{}
-	if start < s.start.Load() {
+	if start < s.Start() {
```

Comment:
We need to be careful about recursive read locks. They cause deadlocks:

> If a goroutine holds a RWMutex for reading and another goroutine might call Lock, no goroutine should expect to be able to acquire a read lock until the initial read lock is released. In particular, this prohibits recursive read locking. This is to ensure that the lock eventually becomes available; a blocked Lock call excludes new readers from acquiring the lock.

https://pkg.go.dev/sync#RWMutex.RLock

I'd suggest having a lowercase `start()` function which doesn't lock the mutex and is used by both `Start()` and this function.

Reply:
Good catch! I have changed it over